### PR TITLE
Replace PAC images in TriggerTemplates

### PR DESCRIPTION
**Pull request by concaf**
# Changes

    Before this commit, the images in TriggerTemplates deployed by Pipelines
    as Code are not substituted with images required for OperatorHub
    deployment.
    
    This commit adds a transformer that replaces images per the environment
    variable `IMAGE_PAC_PIPELINES_AS_CODE`.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
/cc @nikhil-thomas 
Needs to be backported to `release-v0.55.x`

**Comment by vdemeester** [comment]
(ah it's already done in https://github.com/tektoncd/operator/pull/666)

**Comment by concaf** [comment]
@vdemeester #666 is a cherry pick

**Comment by piyush-garg** [comment]
You need to add the prefix here https://github.com/tektoncd/operator/blob/main/pkg/reconciler/common/transformers.go#L38

also add the transformer in reconciler, just adding env will not work

/hold

**Comment by piyush-garg** [comment]
Also you need to write a new transformer, because this is a special case where image is getting replaced in taskspec of a resource of triggertemplate. Current logic is for deployment

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (sm43) [inline pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:102]: is it required to return an error here?
if image not found then we can continue with existing img right?
what happens in development case? when we do `make apply` locally
do we have a image in env?

File: pkg/reconciler/common/transformers.go
Comment (piyush-garg) [inline pkg/reconciler/common/transformers.go:52]: Should we consider this whole, or have a specific format? Like Having Triggertemplate in name so that we know this is for triggertemplates, otherwise if we go with the pattern folowed, this will look like it finds a container with name - `pipelines-as-code` in deployment

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:102]: there can be a case like this env is not present in the development cycle, or in k8s target for now

File: pkg/reconciler/openshift/tektonaddon/transformer.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/transformer.go:171]: I am not able to think of any case making this

File: pkg/reconciler/openshift/tektonaddon/transformer.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/transformer.go:191]: this method with replace any image in taskruns of triggertemplate, we should fins a pattern with name or something that pac image is replaced, this may result in replacing images of any other triggertemplates also

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go]: good catch, fixed in https://github.com/tektoncd/operator/pull/664/commits/1d5be1021ef4faedc26493b48628efb90e6012b8

File: pkg/reconciler/common/transformers.go
Comment (concaf) [reply pkg/reconciler/common/transformers.go:52]: I see how this can be confusing, I'll rename it to `IMAGE_PAC_TRIGGER_TEMPLATES`

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go]: fixed in https://github.com/tektoncd/operator/pull/664/commits/1d5be1021ef4faedc26493b48628efb90e6012b8

File: pkg/reconciler/common/transformers.go
Comment (concaf) [reply pkg/reconciler/common/transformers.go]: Fixed in e1baf385c4f2012d01c33ca0783f29ac912f6d24

File: pkg/reconciler/openshift/tektonaddon/transformer.go
Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:175]: me neither, but I'd rather err on the side of caution here 🙃

File: pkg/reconciler/openshift/tektonaddon/transformer.go
Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:195]: I thought about this as well, but there is no pattern here. I'll ping @chmouel and @sm43 if PAC can guarantee same step names in the future.

Comment (chmouel) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:195]: what kind of guarantee you need? i am not quite understanding the question,

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:195]: @chmouel any pattern that can we identify to replace the image, like step name, etc.
For now, I'll move ahead with what @piyush-garg has suggested.

Comment (nikhil-thomas) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:195]: or it could be any label in the triggerTemplate which this function could use to ensure that the triggertemplate it is working on is from PAC. 

cc @piyush-garg @concaf may be we can keep PAC project as it is and add label through [this](https://github.com/tektoncd/operator/blob/42d68c8428211e993b8455d7de292e0773fc54c9/hack/fetch-releases.sh#L97) if needed.

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:186]: @nikhil-thomas for now, since we don't have any other TriggerTemplate transformer, the current solution should work but going forward we could do something like that, WDYT?

@piyush-garg I've made the changes at 668b77014b8107e0f24ccdda5c25ddbbefb8be71

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/transformer.go:194]: @nikhil-thomas thinking about it, what you said makes sense 😅 added a check here - f43d62ad909f6dc783c22631b98dbaeb9022e465

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:104]: nit: // "triggertemplate_apply_and_launch": "registry.example.io/pac-image"

File: pkg/reconciler/openshift/tektonaddon/pipelinesascode.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: is not better to have this replace on step name inside transformer, so any one using this function has not need to do this every time and also it will be consistent with other transformers

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: which replace are you referring to that happens in other transformers?
https://github.com/concaf/operator/blob/307eafa4a3cce8f01e2354e651c50afb84f3ae93/pkg/reconciler/common/transformers.go#L262
^ `formKey` converts container names to env var names so we can match it, but that's not what I'm doing here. I'm normalizing the step names before so the transformer gets the right step names instead of having to modify it later.

anyway, I'd rather have the transformer get the correct normalized step names as input instead of having the modify and match inside it.

Comment (piyush-garg) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: Yes the idea behind this is, anyone calling this function does not need to do a modification for all calls, rather the transformer is doing the normalization and this is the same thing for all transformers

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: normalization is happening out of the transformer in other calls except the `formKey` 🤔

Comment (piyush-garg) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: I don't have any other strong points behind keeping it inside and outside apart from these two

1. Consistency with the codebase
2. Benefit of having every caller does not need to do the modification

I will let the owners of codebase take the call cc @tektoncd/operator-maintainers

Comment (piyush-garg) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: we can discuss this in WG and can do a follow-up PR for the pattern we agree

Comment (concaf) [reply pkg/reconciler/openshift/tektonaddon/pipelinesascode.go:112]: I think the other transformers are a bit inconsistent in the way of normalizing input names. Let's take an example of `common.DeploymentImages(images)` https://github.com/tektoncd/operator/blob/307eafa4a3cce8f01e2354e651c50afb84f3ae93/pkg/reconciler/common/transformers.go#L177

It takes `images` as input which I'd expect to be a mapping of `container name: image name` and this is true to an extent but not fully.

Some normalization of the container names happens outside the transformer here:
https://github.com/tektoncd/operator/blob/307eafa4a3cce8f01e2354e651c50afb84f3ae93/pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go#L338 where the prefix is removed, lowercased, etc but `_` is not yet replaced by `-` which happens inside the transformer instead - https://github.com/tektoncd/operator/blob/307eafa4a3cce8f01e2354e651c50afb84f3ae93/pkg/reconciler/common/transformers.go#L230

I didn't want this inconsistent normalization to be happening for this transformer and I think it's the caller's responsibility to provide normalized container names and not the transformer's; unless you want to tie the transformer to env variables which I don't think is ideal because it reduces the transformers' scope by quite a bit and it cannot be called for non env var cases.

At some point, we should consider doing normalization outside of other transformers as well 🤔

**Pull request merged**
